src/core/vanilla.ts

```diff
@@ -212,10 +216,21 @@ const setAtomReadPromise = <Value>(
     atom,
     dependencies
   ) as AtomState<Value>
+  delete atomState.i
   atomState.p = promise
   return new Map(wip).set(atom, atomState)
 }
 
+const setAtomInvalidatetd = <Value>(
+  state: State,
+  wip: WorkInProgress,
+  atom: Atom<Value>
+): WorkInProgress => {
+  const atomState = wipAtomState(state, wip, atom) as AtomState<Value>
+  atomState.i = atomState.r
+  return new Map(wip).set(atom, atomState)
```

Comment:
Is `concatMap` better here?

Reply:
It's just one set, so this should be good.